### ISSUE TITLE
Add moreau envelope of functional

### ISSUE THREAD

**Issue by adler-j** [odlgroup/odl#762]
The Moreau envelope of a functional f : X -> R multiplied by a scalar a is given by

env_{a f}(x) = min_{y \in X} {1/(2a) ||x - y||_2^2 + f(y)}

This is basically the function that appears in the proximal. Nicely enough it also has as strong relation to the proximal (and this is why we want to add it). It is always differentiable with gradient:

grad [env_{a f}] = 1 / beta (I - prox[a f])

More information [here](ftp://ftp.math.ucla.edu/pub/camreport/cam12-64.pdf).

Basically adding this would allow us to use functionals that only have a proximal with smooth solvers by approximating them in a nice and uniform way.

**Comment by kohr-h** [comment]
> More information here.

I guess you wanted to include a link.

**Comment by adler-j** [comment]
> I guess you wanted to include a link.

yeah, but I lost it now :/
